cmd/subctl/recover_broker_info.go

```diff
@@ -0,0 +1,148 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subctl
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/subctl/internal/cli"
+	"github.com/submariner-io/subctl/internal/constants"
+	"github.com/submariner-io/subctl/internal/exit"
+	"github.com/submariner-io/subctl/internal/restconfig"
+	"github.com/submariner-io/subctl/pkg/broker"
+	"github.com/submariner-io/subctl/pkg/cluster"
+)
+
+var recoverRestConfigProducer = restconfig.NewProducer().WithPrefixedContext("broker").WithDefaultNamespace(constants.DefaultBrokerNamespace).
+	WithPrefixedNamespace("broker", constants.DefaultBrokerNamespace)
+
+var ipsecFile string
+
+// recoverBrokerInfo represents the reconstruct command.
+var recoverBrokerInfo = &cobra.Command{
+	Use:   "recover-broker-info",
+	Short: "Recovers the broker-info.subm file from the installed Broker",
+	Run: func(cmd *cobra.Command, args []string) {
+		exit.OnError(recoverBroker(cli.NewReporter()))
+	},
+}
+
+func init() {
+	recoverBrokerInfo.PersistentFlags().StringVar(&ipsecFile, "ipsec-psk-from", "",
+		"import IPsec PSK from existing submariner broker file, like broker-info.subm")
+	recoverRestConfigProducer.SetupFlags(recoverBrokerInfo.Flags())
+	rootCmd.AddCommand(recoverBrokerInfo)
+}
+
+func recoverBroker(status reporter.Interface) error {
+	// if --brokerconfig flag provided, get the broker and proceed to get Submariner
+	contextFound, err := recoverRestConfigProducer.RunOnSelectedPrefixedContext("broker",
+		func(brokerCluster *cluster.Info, brokerNamespace string, status reporter.Interface) error {
+			status.Start("Checking if Broker is installed on the specified cluster %s in %s namespace", brokerCluster.Name,
+				brokerNamespace)
+			defer status.End()
+			found, err := brokerCluster.GetBroker(brokerNamespace)
+			if err != nil {
+				return status.Error(err, "error getting Broker")
+			}
+			if found {
+				status.Success("Broker found")
+				status.Start("Checking if there are any cluster(s) joined to the Broker")
+
+				clusters, err := brokerCluster.GetJoinedClusters()
+				if err != nil {
+					return status.Error(err, "error listing joined clusters")
+				}
+
+				if len(clusters) != 0 {
+					status.Success("Found cluster(s) joined to the Broker")
+					status.Start("Checking if Submariner is installed on the same cluster as Broker")
+					for i := range clusters {
+						if clusters[i].Spec.ClusterID == brokerCluster.Name {
+							status.Success("Submariner found")
+							return broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, status, ipsecFile)
```

Comment:
As explained elsewhere, the purpose of searching for the local `Cluster` resource is to handle the case where there are multiple brokers installed, in which case we need to determine if the `Submariner` resource is actually joined to the broker to be recovered. What distinguishes multiple brokers is their namespaces.  So we can simply check if `brokerCluster.Submariner.Spec.BrokerK8sRemoteNamespace == brokerNamespace`.

Reply:
This checks the case where there are multiple clusters, each with broker and Submariner installed.

cluster1                    cluster2
Broker1                    Broker2
Submariner1          Submariner2

**Submariner2 is joined with Broker1. But Submariner1 is not joined with Broker1.**

When running recover command on cluster1 with context = cluster1, it will find both Broker1 and Submariner1. But since Submariner1 is  **NOT** connected to Broker1, it is the wrong Submariner instance for us and hence should not be used. Hence the command will fail saying that "a Submariner cluster" is joined to this Broker but we are not interested in this one. Please specify correct Submariner instance.

Reply:
right so this is the case where the user specified `--brokercontext cluster1`.  In your example, we would find Submariner1 but we need to check if it's joined to the local Broker1. You're searching for the `Cluster` resource but across all namespaces which is problematic - you really should only search the broker namespace. I suggested we check if `brokerCluster.Submariner.Spec.BrokerK8sRemoteNamespace == brokerNamespace` but that's not correct.  I think all we need to do is look up the `Cluster` resource associated with Submariner1, ie

```
cluster := &submarinerv1.Cluster{}
err := c.ClientProducer.ForGeneral().Get(context.TODO(), controllerClient.ObjectKey{
        Namespace: brokerNamespace,
	Name:      brokerCluster.Submariner.Spec.ClusterID,
     }, cluster)
```

If not found then we'd check the cluster associated with `--context` via `RunOnSelectedContext`.  It could be cluster2 or cluster1, either way we need to do the same `Cluster` lookup  as above to ensure it's joined to Broker1.

Reply:
@tpantelis I think the whole confusion can be avoided and the UX can be much simpler - https://github.com/submariner-io/subctl/pull/467#discussion_r1103819587